libretro-common/audio/audio_mixer.c

```diff
@@ -837,6 +844,14 @@ audio_mixer_voice_t* audio_mixer_play(audio_mixer_sound_t* sound,
       if (voice->type != AUDIO_MIXER_TYPE_NONE)
          continue;
 
+      slock_lock(voice->lock);
+
+      if (voice->type != AUDIO_MIXER_TYPE_NONE)
```

Comment:
Why are there two of the same evaluation in here?

`if (voice->type != AUDIO_MIXER_TYPE_NONE)`
should be atomic, just get rid of this one and leave the one above.

Reply:
The one before the lock is not strictly necessary. It's an optimization for the most common case. The one after the lock is always necessary as 2 threads can otherwise acquire the same mixer after checking. Lock has to be held both when checking and when writing with no unlocking in-between. Can we keep the optimization?

Reply:
Yeah, that's fine. Just pointed that out because it seemed redundant.

The way I see it, if "voice->type != AUDIO_MIXER_TYPE_NONE" is often going to be the case, LBYL is going to indeed perform better (as it's right now), otherwise EAFP is likely going to be better (by removing the first evaluation).

I'll leave it up to you and you can mark it as resolved if you think it's best like that.

I also see nothing else that needs changing and I believe it's good to be merged @twinaphex .

Reply:
> Yeah, that's fine. Just pointed that out because it seemed redundant.
> 
> The way I see it, if "voice->type != AUDIO_MIXER_TYPE_NONE" is often going to be the case, LBYL is going to indeed perform better (as it's right now), 

Often first choice is music and so being able to skip it without licking is an advantage.


> otherwise EAFP is likely going to be better (by removing the first evaluation).
> 
> I'll leave it up to you and you can mark it as resolved if you think it's best like that.
> 
> I also see nothing else that needs changing and I believe it's good to be merged @twinaphex .

@twinaphex can we merge this?